File: algoritmo-genetico/individuals.py
from researchers import Researcher
from random import choice
import utils
from datetime import datetime

class Individual:

    # Initializes individual with random departure and returning flights for each researcher
    def __init__(self, flights_organized):
        self.researchers = [
            Researcher(choice(flights_organized["from_LIS"]), choice(flights_organized["to_LIS"])),
            Researcher(choice(flights_organized["from_MAD"]), choice(flights_organized["to_MAD"])),
            Researcher(choice(flights_organized["from_CDG"]), choice(flights_organized["to_CDG"])),
            Researcher(choice(flights_organized["from_DUB"]), choice(flights_organized["to_DUB"])),
            Researcher(choice(flights_organized["from_BRU"]), choice(flights_organized["to_BRU"])),
            Researcher(choice(flights_organized["from_LHR"]), choice(flights_organized["to_LHR"]))
        ]

    def totalFlightCosts(self):
        LIS_cost = self.researchers[0].going_flight.cost + self.researchers[0].returning_flight.cost
        MAD_cost = self.researchers[1].going_flight.cost + self.researchers[1].returning_flight.cost
        CDG_cost = self.researchers[2].going_flight.cost + self.researchers[2].returning_flight.cost
        DUB_cost = self.researchers[3].going_flight.cost + self.researchers[3].returning_flight.cost
        BRU_cost = self.researchers[4].going_flight.cost + self.researchers[4].returning_flight.cost
        LHR_cost = self.researchers[5].going_flight.cost + self.researchers[5].returning_flight.cost
        return LIS_cost + MAD_cost + CDG_cost + DUB_cost + BRU_cost + LHR_cost
    
    def totalAirportWaitingMinutes(self) -> int:
        latest_arrival_flight = utils.getLatestArrivalFlight([
            self.researchers[0].going_flight,
            self.researchers[1].going_flight,
            self.researchers[2].going_flight,
            self.researchers[3].going_flight,
            self.researchers[4].going_flight,
            self.researchers[5].going_flight
        ])

        first_departure_flight = utils.getFirstDepartureFlight([
            self.researchers[0].returning_flight,
            self.researchers[1].returning_flight,
            self.researchers[2].returning_flight,
            self.researchers[3].returning_flight,
            self.researchers[4].returning_flight,
            self.researchers[5].returning_flight
        ])

        total_waiting = 0 # in minutes

        for researcher in self.researchers:
            # waiting on arrival
            latest_time = datetime.strptime(latest_arrival_flight.arrival_time, "%H:%M")
            arrival_time = datetime.strptime(researcher.going_flight.arrival_time, "%H:%M")
            total_waiting = total_waiting + int((latest_time - arrival_time).total_seconds() / 60)

            # waiting for departure
            first_time = datetime.strptime(first_departure_flight.departure_time, "%H:%M")
            departure_time = datetime.strptime(researcher.going_flight.departure_time, "%H:%M")
            total_waiting = total_waiting + int((departure_time - first_time).total_seconds() / 60)

        return total_waiting

    # Returns the correct string for accessing the flight list of interest, based on which researcher and on whether 
    # the departure flight or the returning flight is required ("toOrFromIndex")
    def indexToFlight(self, toOrFromIndex, researcherIndex):

        # 0 = "to" origin country flight (returning flight)
        if toOrFromIndex == 0:
            if researcherIndex == 0:
                return "to_LIS"
            if researcherIndex == 1:
                return "to_MAD"
            if researcherIndex == 2:
                return "to_CDG"
            if researcherIndex == 3:
                return "to_DUB"
            if researcherIndex == 4:
                return "to_BRU"
            if researcherIndex == 5:
                return "to_LHR"
            
        # 1 = "from" origin country flight (departure flight)
        if toOrFromIndex == 1:
            if researcherIndex == 0:
                return "to_LIS"
            if researcherIndex == 1:
                return "to_MAD"
            if researcherIndex == 2:
                return "to_CDG"
            if researcherIndex == 3:
                return "to_DUB"
            if researcherIndex == 4:
                return "to_BRU"
            if researcherIndex == 5:
                return "to_LHR"

    def __str__(self):
        LIS_cost = self.researchers[0].going_flight.cost + self.researchers[0].returning_flight.cost
        MAD_cost = self.researchers[1].going_flight.cost + self.researchers[1].returning_flight.cost
        CDG_cost = self.researchers[2].going_flight.cost + self.researchers[2].returning_flight.cost
        DUB_cost = self.researchers[3].going_flight.cost + self.researchers[3].returning_flight.cost
        BRU_cost = self.researchers[4].going_flight.cost + self.researchers[4].returning_flight.cost
        LHR_cost = self.researchers[5].going_flight.cost + self.researchers[5].returning_flight.cost
        
        return f"LIS researcher flights:\n  Cost: {LIS_cost}\nMAD researcher flights:\n  Cost: {MAD_cost}\nCDG researcher flights:\n  Cost: {CDG_cost}\nDUB researcher flights:\n  Cost: {DUB_cost}\nBRU researcher flights:\n  Cost: {BRU_cost}\nLHR researcher flights:\n  Cost: {LHR_cost}\n"

File: algoritmo-genetico/Golang-version/individuals/individuals.go
package individuals

import (
	"fmt"
	"math/rand"
	"time"

	"ga/flights"
	"ga/researchers"
	"ga/utils"
)

// Individual representa um indivíduo com voos para cada pesquisador
type Individual struct {
	Researchers []researchers.Researcher
}

// NewIndividual cria um novo indivíduo com voos aleatórios para cada pesquisador
func NewIndividual(flightsOrganized map[string][]flights.Flight) Individual {
	return Individual{
		Researchers: []researchers.Researcher{
			{
				GoingFlight:     flightsOrganized["from_LIS"][rand.Intn(len(flightsOrganized["from_LIS"]))],
				ReturningFlight: flightsOrganized["to_LIS"][rand.Intn(len(flightsOrganized["to_LIS"]))],
			},
			{
				GoingFlight:     flightsOrganized["from_MAD"][rand.Intn(len(flightsOrganized["from_MAD"]))],
				ReturningFlight: flightsOrganized["to_MAD"][rand.Intn(len(flightsOrganized["to_MAD"]))],
			},
			{
				GoingFlight:     flightsOrganized["from_CDG"][rand.Intn(len(flightsOrganized["from_CDG"]))],
				ReturningFlight: flightsOrganized["to_CDG"][rand.Intn(len(flightsOrganized["to_CDG"]))],
			},
			{
				GoingFlight:     flightsOrganized["from_DUB"][rand.Intn(len(flightsOrganized["from_DUB"]))],
				ReturningFlight: flightsOrganized["to_DUB"][rand.Intn(len(flightsOrganized["to_DUB"]))],
			},
			{
				GoingFlight:     flightsOrganized["from_BRU"][rand.Intn(len(flightsOrganized["from_BRU"]))],
				ReturningFlight: flightsOrganized["to_BRU"][rand.Intn(len(flightsOrganized["to_BRU"]))],
			},
			{
				GoingFlight:     flightsOrganized["from_LHR"][rand.Intn(len(flightsOrganized["from_LHR"]))],
				ReturningFlight: flightsOrganized["to_LHR"][rand.Intn(len(flightsOrganized["to_LHR"]))],
			},
		},
	}
}

// TotalFlightCosts retorna o custo total dos voos de todos os pesquisadores
func (ind *Individual) TotalFlightCosts() int {
	totalCost := 0
	for _, researcher := range ind.Researchers {
		totalCost += researcher.GoingFlight.Cost + researcher.ReturningFlight.Cost
	}
	return totalCost
}

// TotalAirportWaitingMinutes retorna o tempo total de espera nos aeroportos
func (ind *Individual) TotalAirportWaitingMinutes() int {
	goingFlights := []flights.Flight{
		ind.Researchers[0].GoingFlight,
		ind.Researchers[1].GoingFlight,
		ind.Researchers[2].GoingFlight,
		ind.Researchers[3].GoingFlight,
		ind.Researchers[4].GoingFlight,
		ind.Researchers[5].GoingFlight,
	}

	returningFlights := []flights.Flight{
		ind.Researchers[0].ReturningFlight,
		ind.Researchers[1].ReturningFlight,
		ind.Researchers[2].ReturningFlight,
		ind.Researchers[3].ReturningFlight,
		ind.Researchers[4].ReturningFlight,
		ind.Researchers[5].ReturningFlight,
	}

	latestArrivalFlight := utils.GetLatestArrivalFlight(goingFlights)
	firstDepartureFlight := utils.GetFirstDepartureFlight(returningFlights)

	totalWaiting := 0
	latestTime, _ := time.Parse("15:04", latestArrivalFlight.ArrivalTime)
	firstTime, _ := time.Parse("15:04", firstDepartureFlight.DepartureTime)

	for _, researcher := range ind.Researchers {
		arrivalTime, _ := time.Parse("15:04", researcher.GoingFlight.ArrivalTime)
		totalWaiting += int(latestTime.Sub(arrivalTime).Minutes())

		departureTime, _ := time.Parse("15:04", researcher.ReturningFlight.DepartureTime)
		totalWaiting += int(departureTime.Sub(firstTime).Minutes())
	}

	return totalWaiting
}

// IndexToFlight retorna o nome correto da chave para acessar a lista de voos de interesse
func (ind *Individual) IndexToFlight(toOrFromIndex, researcherIndex int) string {
	if toOrFromIndex == 0 {
		switch researcherIndex {
		case 0:
			return "to_LIS"
		case 1:
			return "to_MAD"
		case 2:
			return "to_CDG"
		case 3:
			return "to_DUB"
		case 4:
			return "to_BRU"
		case 5:
			return "to_LHR"
		}
	} else if toOrFromIndex == 1 {
		switch researcherIndex {
		case 0:
			return "from_LIS"
		case 1:
			return "from_MAD"
		case 2:
			return "from_CDG"
		case 3:
			return "from_DUB"
		case 4:
			return "from_BRU"
		case 5:
			return "from_LHR"
		}
	}

	return ""
}

// String retorna uma representação em string do indivíduo
func (ind *Individual) String() string {
	LISCost := ind.Researchers[0].GoingFlight.Cost + ind.Researchers[0].ReturningFlight.Cost
	MADCost := ind.Researchers[1].GoingFlight.Cost + ind.Researchers[1].ReturningFlight.Cost
	CDGCost := ind.Researchers[2].GoingFlight.Cost + ind.Researchers[2].ReturningFlight.Cost
	DUBCost := ind.Researchers[3].GoingFlight.Cost + ind.Researchers[3].ReturningFlight.Cost
	BRUCost := ind.Researchers[4].GoingFlight.Cost + ind.Researchers[4].ReturningFlight.Cost
	LHRCost := ind.Researchers[5].GoingFlight.Cost + ind.Researchers[5].ReturningFlight.Cost

	return fmt.Sprintf("LIS researcher flights:\n  Cost: %d\nMAD researcher flights:\n  Cost: %d\nCDG researcher flights:\n  Cost: %d\nDUB researcher flights:\n  Cost: %d\nBRU researcher flights:\n  Cost: %d\nLHR researcher flights:\n  Cost: %d\n",
		LISCost, MADCost, CDGCost, DUBCost, BRUCost, LHRCost)
}
